Sources/Regex/AST.swift

```diff
@@ -0,0 +1,44 @@
+/// A regex abstract syntax tree
+public enum AST: Hashable {
+
+  /// ... | ... | ...
+  indirect case alternation([AST])
+
+  /// ... ...
+  indirect case concatenation([AST])
+
+  /// (...)
+  indirect case group(Group, AST)
+
+  /// Group with a registered transform
+  indirect case groupTransform(
+    Group, AST, transform: CaptureTransform)
```

Comment:
CC @rxwei, not sure how you want to best model transforms on the AST. This AST form would be unused by literals probably (unless we do something special with literal syntax).

Reply:
Ideally the pattern DSL should have its own extended AST.  I can refactor this out as when I work on the DSL next.

Reply:
Ah, that makes sense to me. So would regex's AST be one of potentially many children in it? Should the AST contain side-band capture metadata?

Reply:
Yes, I think embedding the regex AST in the more general pattern AST would help us decide whether to use a specialized lowering path for a part of the pattern.  Re. capture metadata, I think yes for each leaf’s type information; the combined type structure can be inferred from the tree.